### ISSUE TITLE
Rule 920100 does not handle CONNECT requests correctly

### ISSUE THREAD

**Issue by RedXanadu** [coreruleset/coreruleset#2220]
### Description

Issue originally reported on StackOverflow, [here](https://stackoverflow.com/questions/69434637/owasp-corerulet-warning-invalid-http-request-line-triggered-by-connect-method/).

Rule 920100 is designed to "Validate request line against the format specified in the HTTP RFC". It includes support for the HTTP `CONNECT` request method. Unfortunately, the rule only recognises a subset of valid `CONNECT` requests.

The rule will allow a `CONNECT` request that uses an IPv4 address on the request line.

The rule will **not** allow a `CONNECT` request that uses a domain on the request line. This causes a false positive.

[RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.6) gives an example of using a domain name in this way, so this behaviour is to be expected and should be allowed and accounted for:

    CONNECT server.example.com:80 HTTP/1.1

Further research and testing required, and then rule 920100 probably needs to be expanded to account for this case.

**Additional question:** ☑
How about IPv6? Is there an RFC allowing for that?

*Answer*: Possibly, but I want to avoid getting stuck figuring out the semantics of RFCs. So, I'm choosing to follow [RFC 2817](https://datatracker.ietf.org/doc/html/rfc2817#section-5.2), which talks about `CONNECT` request lines using a "host name and port number destination… separated by a colon: `CONNECT server.example.com:80 HTTP/1.1`".

**Additional question:** ☑
Is a `t:lowercase` transformation more efficient than a case insensitive regular expression? Could that improve the performance of this rule?

*Answer*: When tested without JIT enabled, it's 54.1 vs 52.1 μs in phase 2 for using `(?i:)` vs `t:lowercase`.

Tested again using the CRS Docker images, and concluded that switching to `t:lowercase` does not yield an improvement. It seems to be quite situational (e.g. one large payload vs many small payloads).

**Additional question:** ☑
Can the regular expression in rule 920100 be disassembled?

*Answer*: Yes! This is now done.

**Useful command:**
Use `cURL` to construct a valid `CONNECT` request for testing:

    curl -v -o/dev/null -p -x localhost:80 http://test.com

### Audit Logs / Triggered Rule Numbers

```
[Tue Oct 05 20:18:38.418643 2021]
        [:error]
        [client 192.168.85.1]
      • ModSecurity: Warning. Match of "rx ^(?i:(?:[a-z]{3,10}\s+(?:\w{3,7}?://[\w\-\./]*(?::\d+)?)?/[^?#]*(?:\?[^#\s]*)?(?:#[\S]*)?|connect (?:\d{1,3}\.){3}\d{1,3}\.?(?::\d+)?|options \*)\s+[\w\./]+|get /[^?#]*(?:\?[^#\s]*)?(?:#[\S]*)?)$" against "REQUEST_LINE" required.
        [file "/opt/httpd-waf/modsecurity.d/rules_3-3-0/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"]
      • [id "920100"]
        [msg "Invalid HTTP Request Line"]
        [data "CONNECT test.com:80 HTTP/1.1"]
        [severity "WARNING"]
        [ver "OWASP_CRS/3.3.0"]
        [tag "paranoia-level/1"]
        [hostname "test.com"]
```

### Your Environment

* CRS version: Tested against v3.4/dev, v3.3.2, v3.3.0
* Paranoia level setting: PL1
* ModSecurity version: Any
* Web Server and version: Any
* Operating System and version: Any

### Confirmation

[X] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by dune73** [comment]
Wow, that was probably the most thorough response in the history of stack overflow. Thank you very much @RedXanadu.
Also for the writeup here.

**Comment by RedXanadu** [comment]
Hi @MichaelJahn2, thanks for posting here!

In my opinion, the issue with rule 920100 is a specific shortcoming. The rule is advertised to "Validate request line[s] against the format specified in the HTTP RFC" and to "…[outline] proper construction for CONNECT… requests."

You've demonstrated that rule 920100 does not currently live up to its description, which is a specific bug we should address.

Would you mind flagging your additional false positives in a separate issue, please? I think it would be easiest to deal with them like that. In a separate issue, they could be looked at it could be decided if rule changes are required, if specific rule exclusions are the correct approach, or something else.

**Comment by RedXanadu** [comment]
## Note

The **existing** regular expression for rule 920100 can be disassembled like so:

```
##! Make all search patterns case insensitive
##!+ i

##! Cover the GET method
^get /[^?#]*(?:\?[^#\s]*)?(?:#[\S]*)?$

##! Cover the CONNECT method
^connect (?:\d{1,3}\.){3}\d{1,3}\.?(?::\d+)?\s+[\w\./]+$

##! Cover the OPTIONS method
^options \*\s+[\w\./]+$

##! Cover other methods of the form METHOD [[scheme]://[host][:port]]/path[?query][#fragment] protocol
##!  Method ---|- Scheme:// -|- Host --|-- Port --| Path |--- Query ---| Fragment | Protocol -|
^[a-z]{3,10}\s+(?:\w{3,7}?://[\w\-\./]*(?::\d+)?)?/[^?#]*(?:\?[^#\s]*)?(?:#[\S]*)?\s+[\w\./]+$
```

This should make rule 920100 much easier to maintain in the future.

This uses the new style regex assembling syntax, so this change can be made once #2223 is merged

**Comment by RedXanadu** [comment]
The new style regex assembling syntax isn't quite merge-ready yet, so this change will need to wait a bit longer.

**Comment by RedXanadu** [comment]
Draft PR #2296 begins to address this. More work needed before it can be considered for merging.

**Comment by RedXanadu** [comment]
While testing, I've noticed that valid `CONNECT` requests trigger rule 920300: "Request Missing an Accept Header" at PL 3.

After a bit of research, it looks like `CONNECT` requests don't include Accept headers, and this is expected and acceptable. 

Will look into adding the `CONNECT` method as exempt from rule 920300.

**Comment by RedXanadu** [comment]
PR #2296 is now ready to go.

**Comment by fzipi** [comment]
Both PRs were merged, closing.

**Comment by MichaelJahn2** [comment]
Interesting: in the field I do not have any issues with 920300 and CONNECT requests. I do however have issues with 920280. I will open another issue regarding this. 

> While testing, I've noticed that valid `CONNECT` requests trigger rule 920300: "Request Missing an Accept Header" at PL 3.
> 
> After a bit of research, it looks like `CONNECT` requests don't include Accept headers, and this is expected and acceptable.

Interesting: in the field I do not have any issues with 920300 and CONNECT requests. I do however have issues with 920280. I will open another issue regarding this.
